Squeezer/src/main/java/uk/org/ngo/squeezer/service/SqueezeService.java

```diff
@@ -654,23 +654,20 @@ private void startForeground() {
 
             ongoingNotification = notificationState();
             NotificationData notificationData = new NotificationData(ongoingNotification);
-            Notification notification;
+            Notification notification = notificationData.builder.build();
 
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-                final MediaMetadataCompat.Builder metaBuilder = new MediaMetadataCompat.Builder();
-                metaBuilder.putString(MediaMetadata.METADATA_KEY_ARTIST, ongoingNotification.artistName);
-                metaBuilder.putString(MediaMetadata.METADATA_KEY_ALBUM, ongoingNotification.albumName);
-                metaBuilder.putString(MediaMetadata.METADATA_KEY_TITLE, ongoingNotification.songName);
-                mMediaSession.setMetadata(metaBuilder.build());
-                notification = notificationData.builder.build();
-            } else {
-                notification = notificationData.builder.build();
-                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
+            if (ongoingNotification != null) {
```

Comment:
This was coming back null some times and causing an NPE later; I put 20 minutes into trying to figure out how it could be null but was unable to figure it out. It seems impossible because it's passed into a constructor before it gets here and that constructor should be throwing an NPE if it was null. Oh well.

Reply:
ongoingNotification is created by notificationState() four lines above. If it's null in this condition I think it's because stopForeground is called (by another thread) immediately after startForeground.
If this is the case it could become null anywhere in startForeground, so the test won't prevent crashes. Also if this is the case we have a similar problem in updateOngoingNotification.
Eg use a local varable (instead of the instance variable) in startForeground.

Reply:
I changed all the members to local where it made sense; I think the only thing I left intact was the optimization where it saved the state to skip some work if it didn't change.

Reply:
beautifully!